### DISCUSSION TITLE
heading towards 4.12.0 release in a week or so

### DISCUSSION THREAD

**Discussion by parrt** [antlr/antlr4#4103] [General]
hi. I need to sneak in a release of antlr within the next 10 days or so and I'm completely occupied with other stuff, but could sneak in some critical fixes and updates if you'd like to get them in. please add @parrt to any comments and PRs so they come to my attention.

There will be lots of fixes to the Go runtime and also the introduction of typescript.

**Comment by kaby76** [comment]
I'm updating the grammars-v4 repo to start testing the TypeScript target in expectation of the release. I'm reading the TypeScript target "readme", but it has an error.
* [/doc/typescript-target.md](https://github.com/antlr/antlr4/blob/bd5a263e933599760abed24162237b92278eb3c2/doc/typescript-target.md?plain=1#L35) says the Antlr tool generates .js files for `-Dlanguage=TypeScript`. I don't get any .js files.
```
$ java -jar /c/Users/Kenne/Documents/GitHub/antlr4-upstream-dev/tool/target/antlr4-4.11.2-SNAPSHOT-complete.jar -Dlanguage=TypeScript
 Arithmetic.g4
02/10-08:06:56 ~/fb
$ ls
Arithmetic.g4      Arithmetic.tokens       ArithmeticLexer.tokens  ArithmeticListener.ts
Arithmetic.interp  ArithmeticLexer.interp  ArithmeticLexer.ts      ArithmeticParser.ts
```

**Comment by KvanTTT** [comment]
I still hope this optimization https://github.com/antlr/antlr4/pull/3676 will be eventually reviewed and merged.

Another fix https://github.com/antlr/antlr4/pull/3626 is also very useful because many users writes ineffective lexers and might encounter performance issues. For instance, this one, quite recent: https://github.com/antlr/antlr4/discussions/4059#discussioncomment-4616162

**Comment by parrt** [comment]
I am going to try to release tomorrow... Hopefully people will be around to answer questions about the various targets if there are problems :)  (e.g., @ericvergnaud, @jimidle )

**Comment by ericvergnaud** [comment]
HeyGreat. I’ll check progress regularly.Envoyé de mon iPhoneLe 18 févr. 2023 à 21:32, Terence Parr ***@***.***> a écrit :﻿
I am going to try to release tomorrow... Hopefully people will be around to answer questions about the various targets if there are problems :)  (e.g., @ericvergnaud, @jimidle )

**Comment by parrt** [comment]
ok, starting process of release. i wonder if tests are green. haha

**Comment by parrt** [comment]
as expected. maven won't test on my machine. yay.

**Comment by parrt** [comment]
and my intellij update wiped out previous config. can't test that way either. yay.

**Comment by parrt** [comment]
ok, intellij building / testing again.

**Comment by parrt** [comment]
Swift failing...
```
org.opentest4j.AssertionFailedError: Test: FullContextIF_THEN_ELSEParse_6; State: Compile; java.lang.Exception: command "swift build -c release -Xswiftc -I/Users/parrt/antlr/code/antlr4/runtime/Swift/.build/release -Xlinker -L/Users/parrt/antlr/code/antlr4/runtime/Swift/.build/release -Xlinker -lAntlr4 -Xlinker -rpath -Xlinker /Users/parrt/antlr/code/antlr4/runtime/Swift/.build/release"
  in /var/folders/w1/_nr4stn13lq0rvjdkwh7q8cc0000gn/T/SwiftRunner-ForkJoinPool-1-worker-31-1676836041187 failed
Cause:
java.lang.InterruptedException: Exit code 1 with output:
Building for production...
remark: Incremental compilation has been disabled: it is not compatible with whole module optimization[1/2] Compiling Test TBaseListener.swift
/private/var/folders/w1/_nr4stn13lq0rvjdkwh7q8cc0000gn/T/SwiftRunner-ForkJoinPool-1-worker-31-1676836041187/main.swift:1:8: error: no such module 'Antlr4'
import Antlr4
       ^
```

**Comment by parrt** [comment]
@ericvergnaud typescript:

```
org.opentest4j.AssertionFailedError: Test: GreedyPositiveClosure; State: Compile; java.lang.Exception: TypeScript ANTLR runtime is not initialized
Test directory: /var/folders/w1/_nr4stn13lq0rvjdkwh7q8cc0000gn/T/TsNodeRunner-ForkJoinPool-1-worker-55-1676836089718
```

**Reply by ericvergnaud** [reply]
All tests pass on GitHub, and on my box. So not sure how to help. Avoid parallelizing because I find it fragilizes the tests.

**Reply by parrt** [reply]
can't run mvn so relying on intellij. Any idea about how to get "TypeScript ANTLR runtime" created?

**Reply by KvanTTT** [reply]
>  Avoid parallelizing because I find it fragilizes the tests.

Please clarify why it fragilizes tests. I fixed problems with parallelization in the previous year and I didn't have problems with them except Swift. But Swift itself works not stable on Windows. Also `TypeScript ANTLR runtime is not initialized` error definitily doesn't relate to parallelization.

**Reply by parrt** [reply]
parallel works well for me

**Comment by parrt** [comment]
btw, i updated:

```
beast:dev:~/antlr/code/antlr4 $ vi /Users/parrt/antlr/code/antlr4/runtime/JavaScript/package-lock.json
beast:dev:~/antlr/code/antlr4 $ vi /Users/parrt/antlr/code/antlr4/runtime/JavaScript/package.json
```

to 4.12.0

**Comment by ericvergnaud** [comment]
Have you run ‘npm run build’ ?  ( as described in the release docs )Envoyé de mon iPhoneLe 19 févr. 2023 à 20:56, Terence Parr ***@***.***> a écrit :﻿
can't run mvn so relying on intellij. Any idea about how to get "TypeScript ANTLR runtime" created?

**Reply by parrt** [reply]
ah no. haven't gotten there yet. just updating version numbers...

npm run build should be part of maven test runs too btw or is this always manual?

**Comment by parrt** [comment]
@ericvergnaud `helpers/package_ts.json` has:

```
  "dependencies": {
    "antlr4": "^4.10.1"
  }
```

should be 4.12.0?

**Comment by ericvergnaud** [comment]
I’ve tried using a ln link to a shared node_modules and that didn’t go well.I’m still using npm link to the antlr runtime and that seems to create problems when running too many tests in parallel Envoyé de mon iPhoneLe 19 févr. 2023 à 20:58, Ivan Kochurkin ***@***.***> a écrit :﻿

Avoid parallelizing because I find it fragilizes the tests.

Please clarify why it fragilizes tests. I fixed problems with parallelization and I didn't have problems with them except Swift. But Swift itself works not stable on Windows.

**Comment by ericvergnaud** [comment]
Yes although it’s overridden by npm linkEnvoyé de mon iPhoneLe 19 févr. 2023 à 21:02, Terence Parr ***@***.***> a écrit :﻿
@ericvergnaud helpers/package_ts.json has:
  "dependencies": {
    "antlr4": "^4.10.1"
  }

should be 4.12.0?

**Reply by parrt** [reply]
ran the build. still same error from intellij. mvn still has its head up its ass haha.

**Comment by ericvergnaud** [comment]
I suspect using a non existing version will failEnvoyé de mon iPhoneLe 19 févr. 2023 à 21:02, Terence Parr ***@***.***> a écrit :﻿
@ericvergnaud helpers/package_ts.json has:
  "dependencies": {
    "antlr4": "^4.10.1"
  }

should be 4.12.0?

**Comment by parrt** [comment]
yay maven:

```
mvn -X release:prepare -Darguments="-DskipTests"
```

```
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for ANTLR 4 4.12.0:
[INFO] 
[INFO] ANTLR 4 ............................................ FAILURE [  2.453 s]
[INFO] ANTLR 4 Runtime .................................... SKIPPED
[INFO] ANTLR 4 Tool ....................................... SKIPPED
[INFO] ANTLR 4 Maven plugin ............................... SKIPPED
[INFO] ANTLR 4 Runtime Tests (4th generation) ............. SKIPPED
[INFO] ANTLR 4 Tool Tests ................................. SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  2.896 s
[INFO] Finished at: 2023-02-19T12:36:47-08:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-release-plugin:3.0.0-M6:prepare (default-cli) on project antlr4-master: Maven execution failed, exit code: '1' -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-release-plugin:3.0.0-M6:prepare (default-cli) on project antlr4-master: Maven execution failed, exit code: '1'
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2 (MojoExecutor.java:347)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:330)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:213)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:175)
    at org.apache.maven.lifecycle.internal.MojoExecutor.access$000 (MojoExecutor.java:76)
...
```

**Reply by parrt** [reply]
deploy works but this next command fails.

**Reply by parrt** [reply]
ah. with that AMAZING error, i discovered it was password issue to maven central.

**Reply by parrt** [reply]
`export GPG_TTY=$(tty)`

**Reply by parrt** [reply]
looks like push might be working...yep, https://oss.sonatype.org/service/local/repositories/releases/content/org/antlr/antlr4-runtime/4.12.0/antlr4-runtime-4.12.0.jar is there.

**Reply by parrt** [reply]
npm publish succeeded, @ericvergnaud !

**Comment by parrt** [comment]
Someone will have to create the windows C# build; can't do from mac

**Reply by kaby76** [reply]
I think the C# library can be built on a Mac because it's dotnet, which builds and runs C# everywhere. But, I don't know if we are still on net6, which is now old.

**Comment by ericvergnaud** [comment]
Currently away from home, I’ll do it end of the weekEnvoyé de mon iPhoneLe 19 févr. 2023 à 22:00, Terence Parr ***@***.***> a écrit :﻿
Someone will have to create the windows C# build; can't do from mac

**Reply by parrt** [reply]
thanks!

**Reply by parrt** [reply]
A C++ for windows would also be useful. I add to website-antlr4/download

**Reply by parrt** [reply]
my windows laptop won't boot. Cost me $800 and got 3 or 4 uses haha.

**Reply by kaby76** [reply]
> my windows laptop won't boot. Cost me $800 and got 3 or 4 uses haha.

I got a Mac that's now a brick cause I wiped the drives (forgot my Apple ID and it's a brick without the ID/password) and Apple won't install on completely empty drives.

**Comment by parrt** [comment]
Ok, done minus windows C# and C++ and then Dart (@lingyv-li).

**Comment by ericvergnaud** [comment]
No there is an issue with code signing and multi target on MacEnvoyé de mon iPhoneLe 19 févr. 2023 à 23:20, Ken Domino ***@***.***> a écrit :﻿
I think the C# library can be built on a Mac because it's dotnet, which builds and runs C# everywhere. But, I don't know if we are still on net6, which is now old.

**Reply by KvanTTT** [reply]
> No there is an issue with code signing and multi target on Mac

It's because csproj contains [`net45` target](https://github.com/antlr/antlr4/blob/master/runtime/CSharp/src/Antlr4.csproj#L7) that is Windows-only.

**Reply by kaby76** [reply]
Can we stop supporting the net45 target? That target is very old, and Microsoft wisely stopped supporting that version seven years ago. As with Net Framework, old releases of Antlr are available. Sorry, but I'm anxious to get the C# nuget package to start testing [grammars-v4 with 4.12](https://github.com/kaby76/grammars-v4/tree/issue-3069) and net7, then start "deep" testing by comparing the parse trees across all targets and input strings. It would be nice to have the PHP target too, but unnecessary because I stopped testing that target for the time being: "grouped parsing", aka "warm-up parsing," which is now the default for testing, does not speed up parsing for PHP.

**Comment by ericvergnaud** [comment]
I’ll release the C# runtime tonight, the below conversation is better held calmly than in emergency

**Reply by kaby76** [reply]
Thanks. Not an emergency, next week will be fine. There are diffs in the output that have to resolve. https://github.com/antlr/antlr4/issues/4132. I will add tree diff checking with every input file across all targets, which I expect will find new bugs. https://github.com/antlr/grammars-v4/issues/2991

**Comment by ericvergnaud** [comment]
Just published 4.12.0 to https://www.nuget.org/packages/Antlr4.Runtime.Standard/

**Reply by parrt** [reply]
yay!

**Comment by ericvergnaud** [comment]
I tested the typescript target (changing version number from the beta), all green

**Comment by ericvergnaud** [comment]
4.12 is out, shall we close this discussion ?

**Reply by parrt** [reply]
yep, didn't know we could "close"...don't see a button for that. i still have to tweet.

**Reply by parrt** [reply]
i "locked"
